### PR TITLE
collapse ipv4 addresses into blocks

### PR DESCRIPTION
**Pull request by jianwan2016**
```
λ> let ips' = (\i -> read i :: I.Ipv4Address) <$> ips
ips' :: [Ipv4Address]
(0.00 secs, 287,928 bytes)
λ> ips'
[104.64.16.1,104.64.16.2,104.64.16.3,104.64.16.4,104.64.16.5,104.64.16.6,104.64.16.7,104.64.16.8,104.64.16.11,104.64.16.12,104.64.16.15,104.64.16.16,104.64.16.17,104.64.16.18,104.64.16.19,104.64.16.20,104.64.16.21,104.64.16.22,104.64.16.23,104.64.16.24,104.64.16.26,104.64.16.29,104.64.16.30,104.64.16.31,104.64.16.32,104.64.16.34,104.64.16.35,104.64.16.37,104.64.16.38,104.64.16.39,104.64.16.40,104.64.16.42,104.64.16.43,104.64.16.44,104.64.16.46,104.64.16.47,104.64.16.48,104.64.16.50,104.64.16.51,104.64.16.53,104.64.16.54,104.64.16.55,104.64.16.57,104.64.16.58,104.64.16.60,104.64.16.61,104.64.16.64,104.64.16.65,104.64.16.67,104.64.16.68,104.64.16.69,104.64.16.71,104.64.16.74,104.64.16.75,104.64.16.78,104.64.16.79,104.64.16.80,104.64.16.81,104.64.16.82,104.64.16.83,104.64.16.84,104.64.16.85,104.64.16.86,104.64.16.87,104.64.16.88,104.64.16.89,104.64.16.90,104.64.16.91,104.64.16.92,104.64.16.93,104.64.16.96,104.64.16.97,104.64.16.98,104.64.16.99,104.64.16.100,104.64.16.102,104.64.16.105,104.64.16.106,104.64.16.107,104.64.16.108,104.64.16.109,104.64.16.110,104.64.16.112,104.64.16.113,104.64.16.114,104.64.16.115,104.64.16.116,104.64.16.118,104.64.16.119,104.64.16.120,104.64.16.122,104.64.16.123,104.64.16.124,104.64.16.125,104.64.16.126,104.64.16.127,104.64.16.128,104.64.16.129,104.64.16.132,104.64.16.134,104.64.16.138,104.64.16.141,104.64.16.142,104.64.16.143,104.64.16.144,104.64.16.146,104.64.16.147,104.64.16.148,104.64.16.149,104.64.16.152,104.64.16.153,104.64.16.154,104.64.16.155,104.64.16.156,104.64.16.158,104.64.16.159,104.64.16.160,104.64.16.161,104.64.16.163,104.64.16.164,104.64.16.166,104.64.16.167,104.64.16.168,104.64.16.170,104.64.16.171,104.64.16.172,104.64.16.173,104.64.16.174,104.64.16.176,104.64.16.178,104.64.16.179,104.64.16.180,104.64.16.181,104.64.16.183,104.64.16.184,104.64.16.185,104.64.16.186,104.64.16.187,104.64.16.189,104.64.16.191,104.64.16.194,104.64.16.195,104.64.16.196,104.64.16.198,104.64.16.199,104.64.16.201,104.64.16.202,104.64.16.203,104.64.16.204,104.64.16.205,104.64.16.206,104.64.16.207,104.64.16.208,104.64.16.209,104.64.16.210,104.64.16.214,104.64.16.215,104.64.16.216,104.64.16.217,104.64.16.218,104.64.16.219,104.64.16.220,104.64.16.222,104.64.16.223,104.64.16.225,104.64.16.226,104.64.16.227,104.64.16.229,104.64.16.230,104.64.16.231,104.64.16.232,104.64.16.233,104.64.16.235,104.64.16.236,104.64.16.238,104.64.16.241,104.64.16.242,104.64.16.244,104.64.16.245,104.64.16.246,104.64.16.247,104.64.16.248,104.64.16.251,104.64.16.252,104.64.16.253,104.64.16.254,104.64.16.255]
it :: [Ipv4Address]
(0.14 secs, 6,641,672 bytes)
λ> let tree = buildIpv4AddressTree ips'
tree :: IPTree
(0.00 secs, 287,800 bytes)
(reverse-i-search)`b': let tree = buildIpv4AddressTree ips'
λ> ipv4AddressesToBlock ips'
[104.64.16.1/32,104.64.16.2/31,104.64.16.4/30,104.64.16.8/32,104.64.16.11/32,104.64.16.12/32,104.64.16.15/32,104.64.16.16/29,104.64.16.24/32,104.64.16.26/32,104.64.16.29/32,104.64.16.30/31,104.64.16.32/32,104.64.16.34/31,104.64.16.37/32,104.64.16.38/31,104.64.16.40/32,104.64.16.42/31,104.64.16.44/32,104.64.16.46/31,104.64.16.48/32,104.64.16.50/31,104.64.16.53/32,104.64.16.54/31,104.64.16.57/32,104.64.16.58/32,104.64.16.60/31,104.64.16.64/31,104.64.16.67/32,104.64.16.68/31,104.64.16.71/32,104.64.16.74/31,104.64.16.78/31,104.64.16.80/29,104.64.16.88/30,104.64.16.92/31,104.64.16.96/30,104.64.16.100/32,104.64.16.102/32,104.64.16.105/32,104.64.16.106/31,104.64.16.108/31,104.64.16.110/32,104.64.16.112/30,104.64.16.116/32,104.64.16.118/31,104.64.16.120/32,104.64.16.122/31,104.64.16.124/30,104.64.16.128/31,104.64.16.132/32,104.64.16.134/32,104.64.16.138/32,104.64.16.141/32,104.64.16.142/31,104.64.16.144/32,104.64.16.146/31,104.64.16.148/31,104.64.16.152/30,104.64.16.156/32,104.64.16.158/31,104.64.16.160/31,104.64.16.163/32,104.64.16.164/32,104.64.16.166/31,104.64.16.168/32,104.64.16.170/31,104.64.16.172/31,104.64.16.174/32,104.64.16.176/32,104.64.16.178/31,104.64.16.180/31,104.64.16.183/32,104.64.16.184/30,104.64.16.189/32,104.64.16.191/32,104.64.16.194/31,104.64.16.196/32,104.64.16.198/31,104.64.16.201/32,104.64.16.202/31,104.64.16.204/30,104.64.16.208/31,104.64.16.210/32,104.64.16.214/31,104.64.16.216/30,104.64.16.220/32,104.64.16.222/31,104.64.16.225/32,104.64.16.226/31,104.64.16.229/32,104.64.16.230/31,104.64.16.232/31,104.64.16.235/32,104.64.16.236/32,104.64.16.238/32,104.64.16.241/32,104.64.16.242/32,104.64.16.244/30,104.64.16.248/32,104.64.16.251/32,104.64.16.252/30]
it :: [Ipv4Block]
(0.11 secs, 5,796,168 bytes)
λ> let blocks = ipv4AddressesToBlock ips'
blocks :: [Ipv4Block]
(0.00 secs, 288,016 bytes)
λ> length blocks
102
it :: Int
(0.01 secs, 4,227,488 bytes)
λ> length ips'
187
it :: Int
(0.00 secs, 288,088 bytes)
```

### CODE REVIEW

**Comment by jianwan2016** [comment]
Weighing:
```
Case                  Allocated        Max        Live  GCs
collapse to blocks  436,263,872  2,228,152  55,484,368  405
```

**Comment by jianwan2016** [comment]
```
Case                  Allocated        Max        Live  GCs
tree                395,601,968  5,791,560  56,813,576  366
collapse to blocks  436,263,872  2,228,152  55,481,720  405
```

**Comment by jianwan2016** [comment]
```Case                                               Allocated        Max        Live  GCs
ip addresses to binary tree                      395,601,968  5,791,560  56,841,128  366
collapsing ip addresses to blocks (binary tree)  436,263,872  2,229,304  55,532,064  405
ip addresses to non-collapsed blocks             300,441,584  1,868,352  11,916,696  290
collapsing blocks (map)                          348,567,128  2,958,920  16,886,080  336
collapsing blocks (mutable hashtable)            358,434,376  3,138,096  23,641,504  343
```

**Comment by jianwan2016** [comment]
```Case                                    Allocated        Max        Live  GCs
ip addresses to non-collapsed blocks  300,441,584  1,868,352  11,905,272  290
collapsing blocks                     340,786,688  1,987,072  17,565,320  329```

**Comment by jianwan2016** [comment]
Collapsing 49M ip addresses into 27M:
```
time stack exec hw-ip
26905492

real	2m0.672s
user	1m56.026s
sys	0m4.074s
```

File: app/Main.hs [hunk 1]
```diff
@@ -1,6 +1,11 @@
 module Main where
 
-import Lib
+import           HaskellWorks.Data.Network.Ip as I
 
 main :: IO ()
-main = someFunc
+main = do
+  let srcFile = "/Users/jwan/whole_internet.txt.blks"
```
Comment (dsturnbull) [inline app/Main.hs]: I guess this will only work on your computer :)

Comment (jianwan2016) [reply app/Main.hs]: I'll remove this, thanks :)

**Pull request closed without merging**